src/utils/pgwire/src/pg_protocol.rs

```diff
@@ -396,21 +396,24 @@ where
         let sql = cstr_to_str(&msg.sql_bytes).unwrap();
         tracing::trace!("(extended query)parse query: {}", sql);
 
-        // Flag indicate whether statement is a query statement.
-        // TODO: regard DML with RETURNING as a query
         let is_query_sql = {
-            let lower_sql = sql.to_ascii_lowercase();
-            lower_sql.starts_with("select")
-                || lower_sql.starts_with("values")
-                || lower_sql.starts_with("show")
-                || lower_sql.starts_with("with")
-                || lower_sql.starts_with("describe")
-                || lower_sql.starts_with("explain")
+            let stmts = Parser::parse_sql(sql)
```

Comment:
Does it mean that we need to parse this statement multiple times? Is there a way to avoid that?

Reply:
I think if we want to avoid multi-parsing SQL times, we need to store the parsed statement in session, and in later messages, the pg client is supposed to send data only rather SQL again.

Reply:
Can we have this info included in `FeParseMessage`?

Reply:
> Does it mean that we need to parse this statement multiple times? Is there a way to avoid that?

Yes. To avoid to parse multiple times, we need to refactor extended query mode in pgwire (as @liurenjie1024 says). This will be complete in #8112.

>I think if we want to avoid multi-parsing SQL times, we need to store the parsed statement in session, and in later messages, the pg client is supposed to send data only rather SQL again.

But this require a big refactor and a new design. This PR is to fix #8089 first.